### PR TITLE
Trie modifications prior to contains

### PR DESCRIPTION
**Pull request by ashtul**
- Add callback for the destruction of payload.
- Add `TrieNode_Get` and `TrieNode_GetValue` to API.

### CODE REVIEW

**Review by oshadmi — COMMENTED** [review]
few comments from interactive discussion

File: src/query.c [hunk 1]
```diff
@@ -490,7 +490,7 @@ static void rangeIterCbStrs(const char *r, size_t n, void *p, void *invidx) {
   rangeItersAddIterator(ctx, ir);
 }
 
-static void rangeIterCb(const rune *r, size_t n, void *p) {
+static int rangeIterCb(const rune *r, size_t n, void *p) {
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/query.c:493]: The callback function type of `TrieRangeCallback` has to be changed for contains

File: src/trie/trie.h [hunk 2]
```diff
@@ -99,6 +101,7 @@ TrieNode *__trie_AddChild(TrieNode *n, const rune *str, t_len offset, t_len len,
 TrieNode *__trie_SplitNode(TrieNode *n, t_len offset);
 
 typedef enum {
+  ADD_IGNORE,
```
Comment (oshadmi) [inline src/trie/trie.h:104]: Not used?

File: src/trie/trie.h [hunk 3]
```diff
@@ -225,4 +234,4 @@ void TrieNode_IterateRange(TrieNode *n, const rune *min, int minlen, bool includ
 #ifdef __cplusplus
 }
 #endif
-#endif
+#endif
```
Comment (oshadmi) [inline src/trie/trie.h:237]: eol

File: src/trie/trie.c [hunk 4]
```diff
@@ -670,24 +693,41 @@ typedef struct {
   void *cbctx;
   bool includeMin;
   bool includeMax;
+
+  // stop if reach limit
+  bool stop;
+
+  // timeout
+  struct timespec timeout;  // milliseconds until timeout
+  size_t timeoutCounter;    // counter to limit number of calls to TimedOut()
```
Comment (oshadmi) [inline src/trie/trie.c:702]: Not used?

**Review by oshadmi — DISMISSED** [review]
👍🏼

File: src/trie/trie.c [hunk 5]
```diff
@@ -3,23 +3,32 @@
...
 }
 
-TriePayload *triePayload_New(const char *payload, uint32_t plen);
-
 // Allocate a new trie payload struct
-inline TriePayload *triePayload_New(const char *payload, uint32_t plen) {
+static inline TriePayload *triePayload_New(const char *payload, uint32_t plen) {
 
   TriePayload *p = rm_malloc(sizeof(TriePayload) + sizeof(char) * (plen + 1));
+  p->data[plen] = '\0';
```
Comment (MeirShpilraien) [inline src/trie/trie.c:18]: This is a bug no?

File: src/trie/trie.c [hunk 6]
```diff
@@ -670,24 +693,37 @@ typedef struct {
...
     // printf("Descending to index %lu\n", ii);
-    rangeIterateSubTree(arr[ii], r);
+    if (rangeIterateSubTree(arr[ii], r) != REDISEARCH_OK) {
+      return REDISEARCH_ERR;
+    }
   }
 
   array_trimm_len(r->buf, array_len(r->buf) - n->len);
+  // char *final_str = runesToStr(r->buf, array_len(r->buf), &len);
+  // printf("%s %s %s\n", before_str, after_str, final_str);
```
Comment (MeirShpilraien) [inline src/trie/trie.c:725]: Delete?

File: src/trie/trie.c (hunk 5 above)
Comment (ashtul) [reply src/trie/trie.c:18]: This is to silence a warning that came up in the cpp tests added.
It NULLs the `+ 1` byte from line 17.

File: tests/cpptests/test_cpp_trie.cpp [hunk 7]
```diff
@@ -121,3 +123,60 @@ TEST_F(TrieTest, testDeepEntry) {
...
   TrieType_Free(t);
 }
+
+/**
+ * This test ensures payload isn't corrupted when the trie changes.
+ */
+TEST_F(TrieTest, testPayload) {
+  char buf1[] = "world";
+  
+  Trie *t = NewTrie(NULL);
```
Comment (MeirShpilraien) [inline tests/cpptests/test_cpp_trie.cpp:133]: Can we add a test that uses the free callback?

**Review by MeirShpilraien — COMMENTED** [review]
Looks good :+1: 
Does the new test cover all the changes?

File: src/trie/trie.c (hunk 5 above)
Comment (MeirShpilraien) [reply src/trie/trie.c:18]: What warning? If we do not really need it to be NULL terminated then why adding it?

Comment (rafie) [reply src/trie/trie.c:18]: Commented out line but still don't see warning.

Comment (ashtul) [reply src/trie/trie.c:18]: The warning is in the sanitizer with the cpp tests running ASSERT_STREQ.

File: tests/cpptests/test_cpp_trie.cpp (hunk 7 above)
Comment (ashtul) [reply tests/cpptests/test_cpp_trie.cpp:133]: Done

**Review by oshadmi — APPROVED** [review]
👍🏼

**Pull request merged**